File: LocalPods/DialogView/DialogView/Elements/ButtonsDialogElement.m
//
//  ButtonsDialogElement.m
//  TCLPlus
//
//  Created by OwenChen on 2020/8/18.
//  Copyright © 2020 TCL Electronics Holdings Ltd. All rights reserved.
//

#import "ButtonsDialogElement.h"

#import "UIFont+Dialog.h"

#define DefaultTextColor @"#000000"          // 默认文本颜色
#define LineColor @"#DDDDDD"                 // 默认线条颜色
static const CGFloat DefaultFontSize = 15.0; // 默认字体大小
static const CGFloat ButtonHeight = 56.0;
static const CGFloat LineHeight = 0.5;
static const NSInteger basicButtonTag = 777;


@interface ButtonsDialogElement ()


@end


@implementation ButtonsDialogElement

#pragma mark - Setter Methods
- (void)setModel:(BasicDialogModel *)model {
    [super setModel:model];

    ButtonsDialogModel *buttonsModel = (ButtonsDialogModel *)self.model;
    switch (buttonsModel.layoutType) {
        case DialogButtonsLayoutHorizontal: {
            [self layoutHorizontalButtons];
        } break;
        case DialogButtonsLayoutVertical: {
            [self layoutVerticalButtons];
        } break;
        default: {
            [self layoutHorizontalButtons];
        } break;
    }
}

- (void)layoutVerticalButtons {
    ButtonsDialogModel *buttonsModel = (ButtonsDialogModel *)self.model;
    CGFloat fontSize = buttonsModel.fontSize > 0 ? buttonsModel.fontSize : DefaultFontSize;

    CGFloat buttonWidth = CGRectGetWidth(self.frame);
    for (NSInteger i = 0; i < buttonsModel.buttonTitles.count; i++) {
        NSString *textColor;
        if (buttonsModel.textColors.count > i) {
            textColor = buttonsModel.textColors[i];
        } else {
            textColor = DefaultTextColor;
        }

        NSString *title = buttonsModel.buttonTitles[i];
        UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
        button.frame = CGRectMake(0, i * ButtonHeight, buttonWidth, ButtonHeight);
        [button setTitle:title forState:UIControlStateNormal];
        [button setTitleColor:[UIColor dialog_colorWithHexString:textColor] forState:UIControlStateNormal];
        [button setTitleColor:[UIColor dialog_colorWithHexString:textColor alpha:0.5] forState:UIControlStateHighlighted];
        [button setTitleColor:[UIColor dialog_colorWithHexString:textColor alpha:0.3] forState:UIControlStateDisabled];
        button.titleLabel.font = buttonsModel.bold ? [UIFont dialog_boldFontWithFontSize:fontSize] : [UIFont dialog_normalFontWithFontSize:fontSize];
        [self addSubview:button];
        button.tag = basicButtonTag + i;

        [button addTarget:self action:@selector(clickButton:) forControlEvents:UIControlEventTouchUpInside];

        if (i != buttonsModel.buttonTitles.count - 1) {
            CALayer *bottomLine = [[CALayer alloc] init];
            bottomLine.frame = CGRectMake(0, floor(ButtonHeight - LineHeight), CGRectGetWidth(button.frame), LineHeight);
            bottomLine.backgroundColor = [UIColor dialog_colorWithHexString:LineColor].CGColor;
            [button.layer addSublayer:bottomLine];

            if (i == 0) {
                CALayer *topLine = [[CALayer alloc] init];
                topLine.frame = CGRectMake(0, 0, CGRectGetWidth(self.frame), LineHeight);
                topLine.backgroundColor = [UIColor dialog_colorWithHexString:LineColor].CGColor;
                [button.layer addSublayer:topLine];
            }
        }
    }
}

- (void)layoutHorizontalButtons {
    ButtonsDialogModel *buttonsModel = (ButtonsDialogModel *)self.model;
    CGFloat fontSize = buttonsModel.fontSize > 0 ? buttonsModel.fontSize : DefaultFontSize;

    CGFloat buttonWidth = CGRectGetWidth(self.frame) / buttonsModel.buttonTitles.count;
    for (NSInteger i = 0; i < buttonsModel.buttonTitles.count; i++) {
        NSString *textColor;
        if (buttonsModel.textColors.count > i) {
            textColor = buttonsModel.textColors[i];
        } else {
            textColor = DefaultTextColor;
        }

        NSString *title = buttonsModel.buttonTitles[i];
        UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
        button.frame = CGRectMake(floor(i * buttonWidth), 0, buttonWidth, ButtonHeight);
        [button setTitle:title forState:UIControlStateNormal];
        [button setTitleColor:[UIColor dialog_colorWithHexString:textColor] forState:UIControlStateNormal];
        [button setTitleColor:[UIColor dialog_colorWithHexString:textColor alpha:0.5] forState:UIControlStateHighlighted];
        [button setTitleColor:[UIColor dialog_colorWithHexString:textColor alpha:0.3] forState:UIControlStateDisabled];
        button.titleLabel.font = buttonsModel.bold ? [UIFont dialog_boldFontWithFontSize:fontSize] : [UIFont dialog_normalFontWithFontSize:fontSize];
        [self addSubview:button];
        button.tag = basicButtonTag + i;

        [button addTarget:self action:@selector(clickButton:) forControlEvents:UIControlEventTouchUpInside];

        if (i != buttonsModel.buttonTitles.count - 1) {
            CALayer *rightLine = [[CALayer alloc] init];
            rightLine.frame = CGRectMake(floor(buttonWidth - LineHeight), 0, LineHeight, ButtonHeight);
            rightLine.backgroundColor = [UIColor dialog_colorWithHexString:LineColor].CGColor;
            [button.layer addSublayer:rightLine];
        }
    }

    CALayer *topLine = [[CALayer alloc] init];
    topLine.frame = CGRectMake(0, 0, CGRectGetWidth(self.frame), LineHeight);
    topLine.backgroundColor = [UIColor dialog_colorWithHexString:LineColor].CGColor;
    [self.layer addSublayer:topLine];
}

#pragma mark - Target Methods
- (void)clickButton:(UIButton *)sender {
    if (self.clickBlock) {
        self.clickBlock(sender.tag - basicButtonTag);
    }
}

#pragma mark - Public Methods
/** 使某个按钮重新恢复点击事件 */
- (void)enableButtonAtIndex:(NSInteger)buttonIndex {
    UIButton *button = [self viewWithTag:buttonIndex + basicButtonTag];
    button.enabled = YES;
}

/**使某个按钮无法点击。*/
- (void)disabledButtonAtIndex:(NSInteger)buttonIndex {
    UIButton *button = [self viewWithTag:buttonIndex + basicButtonTag];
    button.enabled = NO;
}

#pragma mark - Element Height
+ (CGFloat)elementHeightWithModel:(BasicDialogModel *)model elementWidth:(CGFloat)elementWidth {
    ButtonsDialogModel *buttonsModel = (ButtonsDialogModel *)model;
    switch (buttonsModel.layoutType) {
        case DialogButtonsLayoutHorizontal: {
            return ButtonHeight;
        } break;
        case DialogButtonsLayoutVertical: {
            CGFloat height = buttonsModel.buttonTitles.count > 0 ? ButtonHeight * buttonsModel.buttonTitles.count : ButtonHeight;
            return height;
        } break;
        default: {
            return ButtonHeight;
        } break;
    }
}

@end

File: LocalPods/DialogViewSwift/DialogViewSwift/Elements/ButtonsDialogElement.swift
//
//  ButtonsDialogElement.swift
//  TCLHome
//
//  Created by huaizhang.chen on 2022/2/18.
//  Copyright © 2022 TCL Eagle Lab. All rights reserved.
//

import Foundation
import UIKit

class ButtonsDialogElement: BasicDialogElement {
    static let dialogButtonHeight: CGFloat = 56.0

    private let dialogButtonDefaultFontSize: CGFloat = 15.0 // 默认字体大小

    private let dialogButtonLineHeight: CGFloat = 0.5

    private let basicButtonTag: Int = 777

    private let dialogButtonDefaultColor: String = "#000000"

    private let dialogButtonLineColor: String = "#DDDDDD"

    typealias ButtonClick = (_ buttonIndex: Int) -> Void // 点击按钮

    var clickBlock: ButtonClick?

    override func setUpDetailDialogElement() {
        super.setUpDetailDialogElement()
        if let buttonsModel = model as? ButtonsDialogModel {
            switch buttonsModel.layoutType {
            case .horizontal: layoutHorizontalButtons(buttonsModel: buttonsModel)
            case .vertical: layoutVerticalButtons(buttonsModel: buttonsModel)
            }
        }
    }

    private func layoutVerticalButtons(buttonsModel: ButtonsDialogModel) {
        let fontSize: CGFloat = buttonsModel.fontSize > 0.0 ? buttonsModel.fontSize : dialogButtonDefaultFontSize
        let buttonWidth: CGFloat = frame.width

        for (index, text) in buttonsModel.buttonTitles.enumerated() {
            var textColor = ""
            if buttonsModel.textColors.count > index {
                textColor = buttonsModel.textColors[index]
            } else {
                textColor = dialogButtonDefaultColor
            }

            let button = UIButton(type: UIButton.ButtonType.custom)
            button.frame = CGRect(x: 0, y: CGFloat(index) * ButtonsDialogElement.dialogButtonHeight, width: buttonWidth, height: ButtonsDialogElement.dialogButtonHeight)
            button.setTitle(text, for: UIControl.State.normal)
            button.setTitleColor(UIColor(hexString: textColor), for: UIControl.State.normal)
            button.setTitleColor(UIColor(hexString: textColor + "80"), for: UIControl.State.highlighted)
            button.setTitleColor(UIColor(hexString: textColor + "4D"), for: UIControl.State.disabled)
            button.titleLabel?.font = buttonsModel.bold ? UIFont.boldSystemFont(ofSize: fontSize) : UIFont.systemFont(ofSize: fontSize)
            addSubview(button)
            button.tag = basicButtonTag + index

            button.addTarget(self, action: #selector(clickButton(sender:)), for: UIControl.Event.touchUpInside)

            if index != buttonsModel.buttonTitles.count - 1 {
                let bottomLine = CALayer()
                bottomLine.frame = CGRect(x: 0.0, y: floor(ButtonsDialogElement.dialogButtonHeight - dialogButtonLineHeight), width: button.frame.width, height: dialogButtonLineHeight)
                bottomLine.backgroundColor = UIColor(hexString: dialogButtonLineColor)?.cgColor
                button.layer.addSublayer(bottomLine)

                if index == 0 {
                    let topLine = CALayer()
                    topLine.frame = CGRect(x: 0.0, y: 0.0, width: frame.width, height: dialogButtonLineHeight)
                    topLine.backgroundColor = UIColor(hexString: dialogButtonLineColor)?.cgColor
                    button.layer.addSublayer(topLine)
                }
            }
        }
    }

    private func layoutHorizontalButtons(buttonsModel: ButtonsDialogModel) {
        let fontSize: CGFloat = buttonsModel.fontSize > 0.0 ? buttonsModel.fontSize : dialogButtonDefaultFontSize
        let buttonWidth: CGFloat = frame.width / CGFloat(buttonsModel.buttonTitles.count)

        for (index, text) in buttonsModel.buttonTitles.enumerated() {
            var textColor = ""
            if buttonsModel.textColors.count > index {
                textColor = buttonsModel.textColors[index]
            } else {
                textColor = dialogButtonDefaultColor
            }

            let button = UIButton(type: UIButton.ButtonType.custom)
            button.frame = CGRect(x: floor(CGFloat(index) * buttonWidth), y: 0, width: buttonWidth, height: ButtonsDialogElement.dialogButtonHeight)
            button.setTitle(text, for: UIControl.State.normal)
            button.setTitleColor(UIColor(hexString: textColor), for: UIControl.State.normal)
            button.setTitleColor(UIColor(hexString: textColor + "80"), for: UIControl.State.highlighted)
            button.setTitleColor(UIColor(hexString: textColor + "4D"), for: UIControl.State.disabled)
            button.titleLabel?.font = buttonsModel.bold ? UIFont.boldSystemFont(ofSize: fontSize) : UIFont.systemFont(ofSize: fontSize)
            addSubview(button)
            button.tag = basicButtonTag + index

            button.addTarget(self, action: #selector(clickButton(sender:)), for: UIControl.Event.touchUpInside)

            if index != buttonsModel.buttonTitles.count - 1 {
                let rightLine = CALayer()
                rightLine.frame = CGRect(x: floor(buttonWidth - dialogButtonLineHeight), y: 0.0, width: dialogButtonLineHeight, height: ButtonsDialogElement.dialogButtonHeight)
                rightLine.backgroundColor = UIColor(hexString: dialogButtonLineColor)?.cgColor
                button.layer.addSublayer(rightLine)
            }
        }

        let topLine = CALayer()
        topLine.frame = CGRect(x: 0.0, y: 0.0, width: frame.width, height: dialogButtonLineHeight)
        topLine.backgroundColor = UIColor(hexString: dialogButtonLineColor)?.cgColor
        layer.addSublayer(topLine)
    }

    @objc private func clickButton(sender: UIButton) {
        if let block = clickBlock {
            block(sender.tag - basicButtonTag)
        }
    }

    public func enableButton(buttonIndex: Int) {
        let button: UIButton? = viewWithTag(buttonIndex + basicButtonTag) as? UIButton
        button?.isEnabled = true
    }

    public func disableButton(buttonIndex: Int) {
        let button: UIButton? = viewWithTag(buttonIndex + basicButtonTag) as? UIButton
        button?.isEnabled = false
    }

    override class func elementHeight(model: BasicDialogModel, elementWidth: CGFloat) -> CGFloat {
        guard let buttonsModel = model as? ButtonsDialogModel else {
            return 0.0
        }

        switch buttonsModel.layoutType {
        case .horizontal: return dialogButtonHeight
        case .vertical:
            if !buttonsModel.buttonTitles.isEmpty {
                return dialogButtonHeight * CGFloat(buttonsModel.buttonTitles.count)
            } else {
                return dialogButtonHeight
            }
        }
    }
}
